.github/workflows/docs-preview-deploy.yml

```diff
@@ -47,10 +45,10 @@ jobs:
           head_sha: ${{ github.event.workflow_run.head_sha }}
           pull_requests: ${{ tojson(github.event.workflow_run.pull_requests) }}
         run: |
-          PR_NUMBER=$(jq -r '[.[] | select(.head.sha == "${{ env.head_sha }}")][0].number' <<< '${{ env.pull_requests }}')
+          PR_NUMBER=$(jq -r '[.[] | select(.head.sha == "${head_sha}")][0].number' <<< '${pull_requests}')
```

Comment:
I'm not sure what the value of this would be, and I don't think your suggestion here would work as-is due to the use of `'` single-quotes wrapping this expression it would not interpolate the variables. Thus your suggestion would break the intended functionality.

The expression syntax to use [`env` context](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs#example-usage-of-the-env-context) works as that is pre-processed before the `run` executes.

---

Both of these variables are referencing the steps own `env` vars set directly above, which use context that shouldn't be possible to tamper with? Could you please explain the value of this suggestion?

I think there is less risk for error for maintainers with the `env` context expressions used, since it is distinctively clear for us to grok vs the subtle error with quotes in shell scripts affecting interpolation which can introduce bugs as this PR has done unintentionally?

Reply:
The sha is totally safe since its only alphanumeric value. The `github.event.workflow_run.pull_requests` contains untrusted data from the triggering Pull Request though (eg: title and body). An attacker can add a body like ``foo'`id`'bar`` which will pollute the value of the `pull_requests` env var. If you use the `${{}}` expression interpolation, the workflow expressions will get interpolated into the final bash script that will be then passed to bash for exection. So if you use `${{ env.pull_requests }}` the attacker will be able to use the PR body to close the bash single quote and add new commands (eg: ``foo';`id`;bar``).

You are right that `'${pull_requests}'` needs to be replaced with `"${pull_requests}"` since otherwise it wont get expanded. Will fix that in the PR. More info [here](https://securitylab.github.com/resources/github-actions-untrusted-input/)

Reply:
> So if you use `${{ env.pull_requests }}` the attacker will be able to use the PR body to close the bash single quote and add new commands (eg: `` foo';`id`;bar ``).

~~While I get what you're saying here, how is it different from `${pull_requests}`? Both would be a string of JSON key/value pairs?~~

Ah ok, templating to generate script with content vs runtime variable.

> The `github.event.workflow_run.pull_requests` contains untrusted data from the triggering Pull Request though (eg: title and body).

I don't think it does, `pull_request` context has specific metadata like that but not for this `workflow_run` event:

```bash
echo "PRs: ${{ tojson(github.event.workflow_run.pull_requests) }}"
```

```
PRs: [
  {
    base: {
      ref: main,
      repo: {
        id: 506839796,
        name: actions-example,
        url: https://api.github.com/repos/polarathene/actions-example
      },
      sha: e2e0c6555bec8ca661b59057443cfa5a54b8da75
    },
    head: {
      ref: test-branch,
      repo: {
        id: 506839796,
        name: actions-example,
        url: https://api.github.com/repos/polarathene/actions-example
      },
      sha: 113f63be4028584a1528e63ff33f821d990c28d2
    },
    id: 2157169857,
    number: 7,
    url: https://api.github.com/repos/polarathene/actions-example/pulls/7
  }
]
```

No PR title in this data?

Reply:
Good to know, the [docs](https://docs.github.com/en/webhooks/webhook-events-and-payloads#workflow_run) just mention that its an array of pull request objects so I assumed they contained the title and body. I would be in the safe side though, perhaps untrusted data is added to these objects in the future